Use Xcode 15.3 in CI

mokagio opened the pull request:
Builds on top of https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1391, where I noticed that CI didn't run because we no longer have agents with Xcode 13 online.

To test: See green CI running on Xcode 15.3 image... via #1393 which marks four tests as expected failures 😞 

---

- [x] I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary.

mokagio committed:
DRY `env` and `agents` in Buildkite pipeline

mokagio committed:
Address an access control warning

> Non-'@objc' instance method in extensions cannot be overridden;
> use 'public' instead

mokagio committed:
Centralize deployment target definition at projects level

mokagio committed:
Bump deployment target to iOS 12.0, minimum for Xcode 15.3

mokagio committed:
Use Xcode 15.3 in CI

mokagio committed:
Use Ruby 3.2.2 and update RuboCop to compatible version

mokagio committed:
Bump iOS and Simulator versions for CI

mokagio committed:
Allow warnings when validating podspec

Ideally, this should be removed once the warnings have been addressed

mokagio committed:
Update Fastlane in the hope it helps with stuck tests in CI

mokagio committed:
Don't waste time cleaning up Simulators in CI

mokagio commented:
The test that locks is `testCopyAndPasteToPlainText` from `TextViewTests`. It didn't occur for me locally till I tried to run the test class individually.

mokagio commented:
> The test that locks is `testCopyAndPasteToPlainText` from `TextViewTests`. It didn't occur for me locally till I tried to run the test class individually.

Figured out why...

<img width="300" alt="Screenshot 2024-04-16 at 19 49 34" src="https://github.com/wordpress-mobile/AztecEditor-iOS/assets/1218433/b6d41090-6c19-48c5-8def-933896d88374">

mokagio commented:
Ah! I also realized that @jkmassel had already worked out that issue in #1377. I had seen some code about `UIPasteboard` `.forTesting` and was surprised not to find it on my local. I then realized this branch was out of sync with `develop`.

mokagio committed:
Merge remote-tracking branch 'origin/develop' into mokagio/xcode-15.3

mokagio committed:
Merge remote-tracking branch 'origin/tooling/fix-sync-pod-publication' into mokagio/xcode-15.3

AliSoftware committed:
Fix NSSecureCoding implementations

@mokagio This should fix the `testPaste{Image,Video}WithoutFormatting` test failures.
 - I have searched the code for `decodeObject(forKey` occurrences and confirmed there wasn't any left
 - But I have not triple-checked that *every* custom subclass of anything we might encode in our `NSAttributedString` attributes have had `+supportsSecureCoding` redefined

Indeed, apparently, **even if** the parent class already overrides it to return `true`, subclasses which override `init?(coder:)`/`encode(with:)` from their parent class need to also re-override `+supportsSecureCoding`. I've added the overrides in classes that I've modified, but there may be more classes that might not be covered by our unit tests around archiving/pasting but would still require it to be added? So would be worth making another pass to be sure we didn't forget any,

AliSoftware commented:
@mokagio I took the liberty to push a commit to fix the `testPaste{Image,Video}WithoutFormatting` test failures.

 - I have searched the code for `decodeObject(forKey:` occurrences and confirmed that I've replaced all of them with `decodeObject(of: …, forKey: …)` instead
 - But I have not triple-checked that *every* custom subclass of anything we might encode in our `NSAttributedString` attributes have had `+supportsSecureCoding` redefined

Indeed, apparently, **even if** the parent class already overrides it to return `true`, subclasses which override `init?(coder:)`/`encode(with:)` from their parent class need to also re-override `+supportsSecureCoding`.

I've added the overrides in classes that I've modified, but haven't check if maybe there are more classes that might not be covered by our unit tests around `NSAttributedString` archiving/pasting, but would still require it to be added. So would be worth making another pass to be sure we didn't forget any (otherwise the client app would start dismissing those attributes during pasting, while it didn't before our migration to iOS12).

AliSoftware committed:
Use `UTType.utf8PlainText` (explicit encoding) instead of `kUTTypePlainText`

Since newer versions of iOS use this new `public.utf8-plain-text` UTI instead of `public.plain-text` like in previous OS versions.

AliSoftware committed:
Fix overriding non-open class outside module

AliSoftware committed:
Revert "Fix overriding non-open class outside module"

This reverts commit 5c3004b944eed56f286c7034ac02ea9d2716137c, because I'm not convinced that's a good idea.

We should double-check that **not** having `+supportsSecureCoding` declared on this subclass doesn't make it fail to be copy/pasted (i.e. archived/unarchived)—like similar cases happened when running Aztec tests. Or if we need `+supportsSecureCoding`, we need to find a way to override/redefine it across module boundaries…

AliSoftware committed:
Declare `class var supportsSecureCoding` as `open`

AliSoftware commented:
@mokagio Just also pushed https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1392/commits/90c0e4db1246dc4e87f004c4ec96061399447efb to fix the `testCopyAndPasteToPlayText` test failure on extra `\n`.

⚠️  That being said, while the test pass, I'm still not 100% sure why there was a `\n` when we were using `public.plain-text` before, so maybe there's more to it and it could still be worth investigating further (and add more tests to cover the cases we missed if needed)

There's also still another test failing, this time in `WordPressEditor` where `testCaptionShortcodeIsProperlyConvertedIntoFigureTagPreservingNestedTags()` seems to have an extra space in the output. I haven't looked into that one so I'll leave it to you to take a look.

AliSoftware committed:
Simpler code style on UIPasteboard+Helpers > UTType

Commit:
Fix shorcode test by using an attribute with key-value

Commit:
Merge pull request #1395 from wordpress-mobile/fix/shortcode-test

Fix - Shortcode test

mokagio commented:
Closing in favor of #1404
